sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/OnTimerInvokers.java

```diff
@@ -0,0 +1,270 @@
+
+    DoFnSignature signature = DoFnSignatures.INSTANCE.getSignature(fn.getClass());
+
+    try {
+      @SuppressWarnings("unchecked")
+      OnTimerInvoker<InputT, OutputT> invoker =
+          (OnTimerInvoker<InputT, OutputT>)
+              getByteBuddyOnTimerInvokerConstructor(signature, timerId).newInstance(fn);
+      return invoker;
+    } catch (InstantiationException
+        | IllegalAccessException
+        | IllegalArgumentException
+        | InvocationTargetException
+        | SecurityException e) {
+      throw new RuntimeException(
+          String.format(
+              "Unable to bind @%s invoker for %s",
+              DoFn.OnTimer.class.getSimpleName(), fn.getClass().getName()),
+          e);
+    }
+  }
+
+  @Nullable
+  private Constructor<?> getConstructorFor(Class<? extends DoFn<?, ?>> fnClass, String timerId) {
+    Map<String, Constructor<?>> invokerConstructors = constructorCache.get(fnClass);
+    if (invokerConstructors == null) {
+      return null;
+    }
+    return invokerConstructors.get(timerId);
+  }
+
+  private void putConstructor(
+      Class<? extends DoFn<?, ?>> fnClass, String timerId, Constructor<?> constructor) {
+    Map<String, Constructor<?>> invokerConstructors = constructorCache.get(fnClass);
+    if (invokerConstructors == null) {
+      invokerConstructors = new HashMap<>();
+      constructorCache.put(fnClass, invokerConstructors);
+    }
+    invokerConstructors.put(timerId, constructor);
+  }
+
+  /** Returns the constructor for a generated invoker for the requested timer. */
+  private synchronized Constructor<?> getByteBuddyOnTimerInvokerConstructor(
+      DoFnSignature signature, String timerId) {
+
+    Class<? extends DoFn<?, ?>> fnClass = signature.fnClass();
+    Constructor<?> constructor = getConstructorFor(fnClass, timerId);
+    if (constructor == null) {
+      Class<? extends OnTimerInvoker<?, ?>> invokerClass =
+          generateOnTimerInvokerClass(signature, timerId);
+      try {
+        constructor = invokerClass.getConstructor(fnClass);
+      } catch (IllegalArgumentException | NoSuchMethodException | SecurityException e) {
+        throw new RuntimeException(e);
+      }
+      putConstructor(fnClass, timerId, constructor);
+    }
+    return constructor;
+  }
+
+  /**
+   * Generates a {@link OnTimerInvoker} class for the given {@link DoFnSignature} and {@link
+   * TimerId}.
+   */
+  private static Class<? extends OnTimerInvoker<?, ?>> generateOnTimerInvokerClass(
+      DoFnSignature signature, String timerId) {
+    Class<? extends DoFn<?, ?>> fnClass = signature.fnClass();
+
+    final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(fnClass);
+
+    DynamicType.Builder<?> builder =
+        new ByteBuddy()
+            // Create subclasses inside the target class, to have access to
+            // private and package-private bits
+            .with(
+                new NamingStrategy.SuffixingRandom("auxiliary_OnTimer") {
```

Comment:
Put the timer ID into the class name for good measure, sanitizing it against characters disallowed in a name?